libs/gltfio/src/ReusableAnimator.cpp

```diff
@@ -0,0 +1,521 @@
+    return mImpl->animations[animationIndex].name.c_str();
+}
+
+void ReusableAnimator::removeAnimatedAsset() {
+    if (mImpl->assetToAnimate) {
+        for (auto iter : mImpl->zeroTransform) {
+            auto ci = mImpl->transformManager->getInstance(iter.first);
+            mImpl->transformManager->setTransform(ci, iter.second);
+        }
+    }
+
+    mImpl->assetToAnimate = nullptr;
+    mImpl->entityMap.clear();
+    mImpl->zeroTransform.clear();
+    mImpl->renderableManager = nullptr;
+    mImpl->transformManager = nullptr;
+    mImpl->boneMatrices.clear();
+    mImpl->weights.clear();
+
+    if (!mImpl->isMorpherExternal) {
+        if (mImpl->morpher) destroyMorpher(mImpl->morpher);
+    }
+    mImpl->morpher = nullptr;
+}
+
+void ReusableAnimator::addAnimatedAsset(FilamentAsset* assetToAnimate, CPUMorpher* morpher) {
+    FFilamentAsset* asset = upcast(assetToAnimate);
+    mImpl->assetToAnimate = asset;
+    for (auto iter : asset->mNodeMap) {
+        std::string name = iter.first->name;
+        auto entity = iter.second;
+        mImpl->entityMap[name] = entity;
+    }
+    mImpl->renderableManager = &asset->mEngine->getRenderableManager();
+    mImpl->transformManager = &asset->mEngine->getTransformManager();
+    if (morpher) {
+        mImpl->morpher = morpher;
+        mImpl->isMorpherExternal = true;
+    }
+    else {
+        mImpl->morpher = createMorpher(asset);
+        mImpl->isMorpherExternal = false;
+    }
+    
+
+    size_t entityCount = assetToAnimate->getEntityCount();
+    auto* entities = assetToAnimate->getEntities();
+    for (size_t i = 0; i < entityCount; ++i) {
+        auto entity = entities[i];
+        auto instance = mImpl->transformManager->getInstance(entity);
+        mImpl->zeroTransform[entity] = mImpl->transformManager->getTransform(instance);
+    }
+
+    const cgltf_data* srcAsset = mImpl->animationAsset->mSourceAsset->hierarchy;
+    const cgltf_animation* srcAnims = srcAsset->animations;
+    for (cgltf_size i = 0, len = srcAsset->animations_count; i < len; ++i) {
+        const cgltf_animation& srcAnim = srcAnims[i];
+        Animation& dstAnim = mImpl->animations[i];
+        mImpl->addChannels(mImpl->entityMap, srcAnim, dstAnim);
+    }
+}
+
+CPUMorpher* ReusableAnimator::createMorpher(FilamentAsset* asset) {
+    return new CPUMorpher(upcast(asset));
+}
+
+void ReusableAnimator::destroyMorpher(CPUMorpher* morpher) {
+    if (morpher) delete morpher;
+}
+
+
+void ReusableAnimatorImpl::addChannels(const NamedEntityMap& entityMap, const cgltf_animation& srcAnim,
+        Animation& dst) {
+    cgltf_animation_channel* srcChannels = srcAnim.channels;
+    cgltf_animation_sampler* srcSamplers = srcAnim.samplers;
+    const Sampler* samplers = dst.samplers.data();
+    for (cgltf_size j = 0, nchans = srcAnim.channels_count; j < nchans; ++j) {
+        const cgltf_animation_channel& srcChannel = srcChannels[j];
+        auto iter = entityMap.find(srcChannel.target_node->name);
```

Comment:
This is the only major logic change we need to get animator to work on a different asset. Just need to find entity by node name instead of cgltf pointer.